manta-crypto/src/accumulator.rs

```diff
@@ -212,129 +220,36 @@ pub trait OptimizedAccumulator: Accumulator {
     }
 }
 
-/// Item Hash Accumulator Model
-pub struct ItemHashAccumulatorModel<T, H, A>
-where
-    H: ItemHashFunction<T>,
-    A: Accumulator<Item = H::Item>,
-{
-    /// Item Hash Function
-    item_hash_function: H,
-
-    /// Accumulator
-    accumulator: A,
-
-    /// Type Parameter Marker
-    __: PhantomData<T>,
-}
-
-impl<T, H, A> Types for ItemHashAccumulatorModel<T, H, A>
-where
-    H: ItemHashFunction<T>,
-    A: Accumulator<Item = H::Item>,
-{
-    type Item = T;
-    type Witness = Witness<A>;
-    type Output = Output<A>;
-}
-
-impl<T, H, A> Model for ItemHashAccumulatorModel<T, H, A>
-where
-    H: ItemHashFunction<T>,
-    A: Accumulator<Item = H::Item>,
-{
-    type Verification = <A::Model as Model>::Verification;
-
-    #[inline]
-    fn verify(
-        &self,
-        item: &Self::Item,
-        witness: &Self::Witness,
-        output: &Self::Output,
-        compiler: &mut (),
-    ) -> Self::Verification {
-        self.accumulator.model().verify(
-            &self.item_hash_function.item_hash(item, compiler),
-            witness,
-            output,
-            compiler,
-        )
-    }
-}
-
-/// Item Hash Accumulator
-pub struct ItemHashAccumulator<T, H, A>
-where
-    H: ItemHashFunction<T>,
-    A: Accumulator<Item = H::Item>,
-{
-    /// Item Hash Accumulator Model
-    model: ItemHashAccumulatorModel<T, H, A>,
-}
-
-impl<T, H, A> Accumulator for ItemHashAccumulator<T, H, A>
-where
-    H: ItemHashFunction<T>,
-    A: Accumulator<Item = H::Item>,
-{
-    type Item = T;
-    type Model = ItemHashAccumulatorModel<T, H, A>;
-
-    #[inline]
-    fn model(&self) -> &Self::Model {
-        &self.model
-    }
-
-    #[inline]
-    fn insert(&mut self, item: &Self::Item) -> bool {
-        self.model
-            .accumulator
-            .insert(&self.model.item_hash_function.item_hash(item, &mut ()))
-    }
-
-    #[inline]
-    fn prove(&self, item: &Self::Item) -> Option<MembershipProof<Self::Model>> {
-        self.model
-            .accumulator
-            .prove(&self.model.item_hash_function.item_hash(item, &mut ()))
-            .map(MembershipProof::into)
-    }
-
-    #[inline]
-    fn contains(&self, item: &Self::Item) -> bool {
-        self.model
-            .accumulator
-            .contains(&self.model.item_hash_function.item_hash(item, &mut ()))
-    }
-}
-
 /// Accumulator Membership Proof
-pub struct MembershipProof<M, COM = ()>
+#[derive(derivative::Derivative)]
+#[derivative(
+    Clone(bound = "M::Witness: Clone, M::Output: Clone"),
+    Copy(bound = "M::Witness: Copy, M::Output: Copy"),
+    Debug(bound = "M::Witness: Debug, M::Output: Debug"),
+    Default(bound = "M::Witness: Default, M::Output: Default"),
+    Eq(bound = "M::Witness: Eq, M::Output: Eq"),
+    Hash(bound = "M::Witness: Hash, M::Output: Hash"),
+    PartialEq(bound = "M::Witness: PartialEq, M::Output: PartialEq")
+)]
+pub struct MembershipProof<M>
 where
-    M: Model<COM> + ?Sized,
+    M: Types + ?Sized,
```

Comment:
Why we put `M::Witness: Clone` in `Clone(bound = ...)` instead of having `witness: M::Witness + Clone`?

Reply:
Not sure what this means. `witness: M::Witness` is a declaration of the field of a struct, you cannot put a trait here. The `Clone` bound is saying that if the `MembershipProof` type implements `Clone` it is because `M::Witness: Clone`.